bindings/serialize.cc

```diff
@@ -0,0 +1,405 @@
+  this->id = id;
+  this->start = start;
+  this->limit = limit;
+  this->offset = offset;
+  this->fileX = fileX;
+  this->buildIDX = buildIDX;
+  this->hasFunctions = hasFunctions;
+  this->hasFilenames = hasFilenames;
+  this->hasLineNumbers = hasLineNumbers;
+  this->hasInlineFrames = hasInlineFrames;
+}
+void Mapping::encode(std::vector<char> &b) {
+  encodeUint64Opt(b, 1, id);
+  encodeUint64Opt(b, 2, start);
+  encodeUint64Opt(b, 3, limit);
+  encodeUint64Opt(b, 4, offset);
+  encodeInt64Opt(b, 5, fileX);
+  encodeInt64Opt(b, 6, buildIDX);
+  encodeBoolOpt(b, 7, hasFunctions);
+  encodeBoolOpt(b, 8, hasFilenames);
+  encodeBoolOpt(b, 9, hasLineNumbers);
+  encodeBoolOpt(b, 10, hasInlineFrames);
+}
+
+Line::Line(uint64_t functionID, int64_t line) {
+  this->functionID = functionID;
+  this->line = line;
+}
+void Line::encode(std::vector<char> &b) {
+  encodeUint64Opt(b, 1, functionID);
+  encodeInt64Opt(b, 2, line);
+}
+
+ProfileFunction::ProfileFunction(uint64_t id, int64_t nameX,
+                                 int64_t systemNameX, int64_t filenameX,
+                                 int64_t startLine) {
+  this->id = id;
+  this->nameX = nameX;
+  this->systemNameX = systemNameX;
+  this->filenameX = filenameX;
+  this->startLine = startLine;
+}
+void ProfileFunction::encode(std::vector<char> &b) {
+  encodeUint64Opt(b, 1, id);
+  encodeInt64Opt(b, 2, nameX);
+  encodeInt64Opt(b, 3, systemNameX);
+  encodeInt64Opt(b, 4, filenameX);
+  encodeInt64Opt(b, 5, startLine);
+}
+
+ProfileLocation::ProfileLocation(uint64_t id, uint64_t mappingID,
+                                 uint64_t address, std::vector<Line> line,
+                                 bool isFolded) {
+  this->id = id;
+  this->mappingID = mappingID;
+  this->address = address;
+  this->line = line;
+  this->isFolded = isFolded;
+}
+void ProfileLocation::encode(std::vector<char> &b) {
+  encodeUint64Opt(b, 1, id);
+  encodeInt64Opt(b, 2, mappingID);
+  encodeInt64Opt(b, 3, address);
+  for (std::vector<Line>::iterator l = line.begin(); l != line.end(); ++l) {
+    encodeMessage(b, 4, &*l);
+  }
+  encodeBoolOpt(b, 5, isFolded);
+}
+
+Sample::Sample(std::vector<uint64_t> locationID, std::vector<int64_t> value,
+               std::vector<Label> label) {
+  this->locationID = locationID;
+  this->value = value;
+  this->label = label;
+}
+void Sample::encode(std::vector<char> &b) {
+  encodeUint64s(b, 1, locationID);
+  encodeInt64s(b, 2, value);
+  for (std::vector<Label>::iterator l = label.begin(); l != label.end(); ++l) {
+    encodeMessage(b, 3, &*l);
+  }
+}
+
+Profile::Profile(std::string periodType, std::string periodUnit, int64_t period,
+                 int64_t timeNanos, int64_t durationNanos,
+                 std::string dropFrames, std::string keepFramesX) {
+  // first index of strings must be ""
+  getStringId("");
+
+  this->periodType =
+      ValueType(getStringId(periodType), getStringId(periodUnit));
+  this->period = period;
```

Comment:
To initialize fields in the constructor it's more common to use the initializer list like

```
Profile::Profile(std::string periodType, std::string periodUnit, int64_t period,
￼                 int64_t timeNanos, int64_t durationNanos,
￼                 std::string dropFrames, std::string keepFramesX) : period(period) {
```